c10/xpu/XPUFunctions.cpp

```diff
@@ -35,11 +53,25 @@ void enumDevices(std::vector<std::unique_ptr<sycl::device>>& devices) {
       continue;
     }
     auto device_list = platform.get_devices();
-    for (const auto& device : device_list) {
-      if (device.is_gpu()) {
-        devices.push_back(std::make_unique<sycl::device>(device));
+    auto is_igpu = std::vector<bool>(device_list.size(), false);
+    auto is_dgpu = std::vector<bool>(device_list.size(), false);
+    identifyDeviceType(device_list, is_igpu, is_dgpu);
+    // First, enumerate dGPU devices.
+    for (const auto i : c10::irange(device_list.size())) {
+      if (is_dgpu[i]) {
+        devices.push_back(std::make_unique<sycl::device>(device_list[i]));
+      }
+    }
+    // Second, enumerate iGPU devices if no dGPU is found.
+    if (devices.empty()) {
+      for (const auto i : c10::irange(device_list.size())) {
+        if (is_igpu[i]) {
+          devices.push_back(std::make_unique<sycl::device>(device_list[i]));
+        }
```

Comment:
By now, we can only enumerate all the devices of the first platform. Otherwise, the logic is a mess. @gujinghui , any comments?

Reply:
Sounds like current code is not correct. **Without** driver fix, our case should be,
1. if only iGPU in single platform, let's take iGPU.
1. If all GPUs in single platform, we should filter out the iGPU only, and take all dGPUs in the platform.
2. If GPUs located in multi platforms, we should only take the **first** platform **without** iGPU.

The logic is not trivial...

Reply:
I think the more intricate logic will make XPU harder to communicate with a third-party library.

Reply:
The update code follows @gujinghui logic.